identity_stardust/src/did/stardust_did.rs

```diff
@@ -193,18 +217,55 @@ impl StardustDID {
       // Self::DEFAULT_NETWORK is built from a static reference so unwrapping is fine
       .unwrap_or((Self::DEFAULT_NETWORK, input))
   }
+}
 
-  /// Returns the unique tag of the `DID`.
-  pub fn tag(&self) -> &str {
-    Self::denormalized_components(self.method_id()).1
+#[cfg(feature = "iota-client")]
+mod stardust_did_iota_client {
+  use std::ops::Deref;
+
+  use iota_client::block::output::AliasId;
+  use iota_client::block::output::Output;
+  use iota_client::block::output::OutputId;
+  use iota_client::block::payload::transaction::TransactionEssence;
+  use iota_client::block::payload::Payload;
+  use iota_client::block::Block;
+
+  use super::*;
+
+  impl StardustDID {
+    /// Extracts an Alias ID from a published block and returns a new DID from it.
+    ///
+    /// NOTE: if there are multiple Alias Outputs in the payload, this uses the first one.
+    // TODO: can hopefully remove from public API if the publishing logic is wrapped.
+    pub fn from_block(block: &Block, network: &NetworkName) -> crate::Result<StardustDID> {
+      let payload: &Payload = block.payload().ok_or(DIDError::Other("block missing payload"))?;
+      // TODO: improve error handling messages.
+      let output_id: OutputId = get_alias_output_id_from_payload(payload)?;
+      let id: AliasId = AliasId::from(output_id);
+      Ok(StardustDID::new(id.deref(), network))
+    }
   }
 
-  /// Replace the network name of this [`StardustDID`] leaving all other segments (did, method, tag) intact.  
-  pub fn with_network_name(mut self, name: NetworkName) -> Self {
-    let new_method_id: String = format!("{}:{}", name, self.tag());
-    // unwrap is fine as we are only replacing the network
-    self.0.set_method_id(new_method_id).unwrap();
-    self
+  // helper function to get the output id for the first alias output
+  // TODO: improve error handling
+  fn get_alias_output_id_from_payload(payload: &Payload) -> crate::Result<OutputId> {
+    match payload {
+      Payload::Transaction(tx_payload) => {
+        let TransactionEssence::Regular(regular) = tx_payload.essence();
+        for (index, output) in regular.outputs().iter().enumerate() {
+          if let Output::Alias(_) = output {
+            return Ok(OutputId::new(
+              tx_payload.id(),
+              index
+                .try_into()
+                .map_err(|_| DIDError::Other("output index exceeds u16"))?,
+            )?);
+          }
+        }
+        Err(DIDError::Other("no alias output in transaction essence"))?
+      }
+      _ => Err(DIDError::Other("not a transaction payload"))?,
+    }
```

Comment:
nit: I find this more readable. Optional. 

```suggestion 
    if let Payload::Transaction(tx_payload) = payload {
      let TransactionEssence::Regular(regular) = tx_payload.essence();
      if let Some((index, _)) = regular
        .outputs()
        .iter()
        .enumerate()
        .find(|(_, item)| matches!(item, Output::Alias(_)))
      {
        Ok(OutputId::new(
          tx_payload.id(),
          index
            .try_into()
            .map_err(|_| DIDError::Other("output index exceeds u16"))?,
        )?)
      } else {
        Err(DIDError::Other("no alias output in transaction essence"))?
      }
    } else {
      Err(DIDError::Other("not a transaction payload"))?
    }
```